src/transport/SecureSessionMgr.cpp

```diff
@@ -110,55 +127,94 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
     mPeerConnections.MarkConnectionActive(state);
 
     {
-        uint8_t * data = nullptr;
+        uint8_t * data         = nullptr;
+        uint8_t * p            = nullptr;
+        uint32_t msgId         = 0;
+        uint32_t payloadLength = 0; // Make sure it's big enough to add two 16-bit ints without overflowing.
+        uint16_t len           = 0;
         PacketHeader packetHeader;
         MessageAuthenticationCode mac;
 
         const uint16_t headerSize = payloadHeader.EncodeSizeBytes();
         uint16_t actualEncodedHeaderSize;
         uint16_t totalLen = 0;
         uint16_t taglen   = 0;
-        uint32_t payloadLength; // Make sure it's big enough to add two 16-bit
-                                // ints without overflowing.
-        static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
-                      "Addition to generate payloadLength might overflow");
-        payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
-        VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+        if (!isEncrypted)
+        {
+            msgId = state->GetSendMessageIndex();
+
+            static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                          "Addition to generate payloadLength might overflow");
+            payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+            VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+        }
+        else
+        {
+            msgId         = msgIdIn;
+            payloadLength = payloadLenIn;
+        }
 
         packetHeader
             .SetSourceNodeId(mLocalNodeId)              //
             .SetDestinationNodeId(peerNodeId)           //
-            .SetMessageId(state->GetSendMessageIndex()) //
+            .SetMessageId(msgId)                        //
```

Comment:
The packet header (unencrypted part of the message header) is used as AAD for message integrity check calculation. The integrity check is calculated while the message is being encrypted. Since we are not re-encrypting the previously encrypted message (when `isEncrypted` is `True`), this particular code is not useful. Probably better to read the packet header directly from the encrypted message buffer, instead of calculating it here.

Reply:
Done
1. Rewind the packet buffer to point to the beginning of packet buffer instead of encrypted paylaod.
2. When the encrypted packet is being sent, extract the packet header from it, and advance the buffer to point past the unencrypted header.

Reply:
> @yufengwangca , thanks for addressing the comments. There's a minor comment. Otherwise, it looks good to me.

Hi Pankaj,

From my stress testing, I found the current solution is far less robust than the previous solution, SecureSessionMgr fails to decode the packet header randomly.  CI also randomly failed because of the same issue. 

I think this is caused by the following tricky logic changes to reuse packet header:
1. Hard rewind the packet Start to the beginning of packet header after each send
2. In the following encrypted packet resend, extract the packet header from it, and advance the buffer to point past the unencrypted header.
3. UDP re-attach the decoded packet header on top of it.

The previous solution always re-calculate the packet header from a clean state which is logically much simpler, and robust in testing. 

I would suggest proceed with the previous solution for now and use a follow-up ticket to investigate the current solution until we can make it stable